File: python/lib/slack_notifier.py
#!/usr/bin/env python3
"""
Slack notification manager for the Issue Orchestrator.

This module handles sending notifications to Slack about issue processing
status, errors, and completion.
"""

import logging
from slack_sdk.web.async_client import AsyncWebClient
from slack_sdk.errors import SlackApiError


class SlackNotifier:
    """Slack notification manager"""
    
    def __init__(self, bot_token: str, channel_id: str):
        self.client = AsyncWebClient(token=bot_token)
        self.channel_id = channel_id
    
    async def send_message(self, text: str = None, blocks: list = None):
        """Send a message to Slack"""
        try:
            # Send message with text or blocks
            response = await self.client.chat_postMessage(
                channel=self.channel_id,
                text=text,
                blocks=blocks
            )
            return response
        except SlackApiError as e:
            logging.error(f"Failed to send Slack message: {e.response['error']}")
    
    async def notify_start(self, issue_number: int, title: str, repo_name: str = None):
        """Notify that processing has started"""
        repo_text = f" in {repo_name}" if repo_name else ""
        issue_url = f"https://github.com/{repo_name}/issues/{issue_number}" if repo_name else f"#{issue_number}"
        issue_link = f"<{issue_url}|#{issue_number}>" if repo_name else f"#{issue_number}"
        
        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f":rocket: *Started processing issue {issue_link}{repo_text}*\n{title}"
                }
            }
        ]
        await self.send_message(
            text=f"Started issue #{issue_number}: {title}",
            blocks=blocks
        )
    
    async def notify_complete(self, issue_number: int, duration: str, repo_name: str = None):
        """Notify that processing is complete"""
        repo_text = f" in {repo_name}" if repo_name else ""
        issue_url = f"https://github.com/{repo_name}/issues/{issue_number}" if repo_name else f"#{issue_number}"
        issue_link = f"<{issue_url}|#{issue_number}>" if repo_name else f"#{issue_number}"
        
        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f":white_check_mark: *Completed issue {issue_link}{repo_text}*\nDuration: `{duration}`"
                }
            }
        ]
        await self.send_message(
            text=f"Completed issue #{issue_number} [{duration}]",
            blocks=blocks
        )
    
    async def notify_needs_input(self, issue_number: int, output: str, repo_name: str = None):
        """Notify that Claude needs input"""
        repo_text = f" in {repo_name}" if repo_name else ""
        issue_url = f"https://github.com/{repo_name}/issues/{issue_number}" if repo_name else f"#{issue_number}"
        issue_link = f"<{issue_url}|#{issue_number}>" if repo_name else f"#{issue_number}"
        
        # Truncate output for Slack
        truncated_output = output[-500:] if len(output) > 500 else output
        
        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f":hourglass: *Issue {issue_link}{repo_text} needs input*"
                }
            },
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"```{truncated_output}```"
                }
            }
        ]
        await self.send_message(
            text=f"Issue #{issue_number} needs input",
            blocks=blocks
        )
    
    async def notify_error(self, issue_number: int, error: str, output: str = None, repo_name: str = None):
        """Notify about an error"""
        repo_text = f" in {repo_name}" if repo_name else ""
        issue_url = f"https://github.com/{repo_name}/issues/{issue_number}" if repo_name else f"#{issue_number}"
        issue_link = f"<{issue_url}|#{issue_number}>" if repo_name else f"#{issue_number}"
        
        blocks = [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f":x: *Error on issue {issue_link}{repo_text}*\n{error}"
                }
            }
        ]
        
        if output:
            # Truncate output for Slack
            truncated_output = output[-300:] if len(output) > 300 else output
            blocks.append({
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"*Last output:*\n```{truncated_output}```"
                }
            })
        
        await self.send_message(
            text=f"Error on issue #{issue_number}: {error}",
            blocks=blocks
        )

File: src/notifiers/slackNotifier.ts
import { formatCompletionMessage } from "./templates";

interface SlackMessage {
  text?: string;
  blocks?: unknown[];
}

export class SlackNotifier {
  private readonly clientToken: string;
  private readonly channelId: string;

  constructor(botToken: string, channelId: string) {
    this.clientToken = botToken;
    this.channelId = channelId;
  }

  private async sendMessage(payload: SlackMessage): Promise<void> {
    if (!this.clientToken || !this.channelId) return;

    const response = await fetch("https://slack.com/api/chat.postMessage", {
      method: "POST",
      headers: {
        "Content-Type": "application/json; charset=utf-8",
        Authorization: `Bearer ${this.clientToken}`,
      },
      body: JSON.stringify({ channel: this.channelId, ...payload }),
    });

    if (!response.ok) {
      console.error(`Failed to send Slack message: ${response.status}`);
    } else {
      const data = await response.json();
      if (!data.ok) {
        console.error(`Slack API error: ${data.error}`);
      }
    }
  }

  async notifyStart(issueNumber: number, title: string, repoName?: string): Promise<void> {
    const repoText = repoName ? ` in ${repoName}` : "";
    const issueUrl = repoName ? `https://github.com/${repoName}/issues/${issueNumber}` : `#${issueNumber}`;
    const issueLink = repoName ? `<${issueUrl}|#${issueNumber}>` : `#${issueNumber}`;

    await this.sendMessage({
      text: `Started issue #${issueNumber}: ${title}`,
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: `:rocket: *Started processing issue ${issueLink}${repoText}*\n${title}`,
          },
        },
      ],
    });
  }

  async notifyComplete(
    issueNumber: number,
    duration: string,
    processorName: string,
    repoName?: string
  ): Promise<void> {
    const message = formatCompletionMessage(issueNumber, duration, processorName, repoName);
    await this.sendMessage({
      text: message,
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: message,
          },
        },
      ],
    });
  }

  async notifyNeedsInput(issueNumber: number, output: string, repoName?: string): Promise<void> {
    const repoText = repoName ? ` in ${repoName}` : "";
    const issueUrl = repoName ? `https://github.com/${repoName}/issues/${issueNumber}` : `#${issueNumber}`;
    const issueLink = repoName ? `<${issueUrl}|#${issueNumber}>` : `#${issueNumber}`;
    const snippet = output.length > 500 ? output.slice(-500) : output;

    await this.sendMessage({
      text: `Issue #${issueNumber} needs input`,
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: `:hourglass: *Issue ${issueLink}${repoText} needs input*`,
          },
        },
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: `\`\`\`${snippet}\`\`\``,
          },
        },
      ],
    });
  }

  async notifyError(issueNumber: number, error: string, output?: string, repoName?: string): Promise<void> {
    const repoText = repoName ? ` in ${repoName}` : "";
    const issueUrl = repoName ? `https://github.com/${repoName}/issues/${issueNumber}` : `#${issueNumber}`;
    const issueLink = repoName ? `<${issueUrl}|#${issueNumber}>` : `#${issueNumber}`;
    const blocks: any[] = [
      {
        type: "section",
        text: {
          type: "mrkdwn",
          text: `:x: *Error on issue ${issueLink}${repoText}*\n${error}`,
        },
      },
    ];

    if (output) {
      const snippet = output.length > 300 ? output.slice(-300) : output;
      blocks.push({
        type: "section",
        text: {
          type: "mrkdwn",
          text: `*Last output:*\n\`\`\`${snippet}\`\`\``,
        },
      });
    }

    await this.sendMessage({
      text: `Error on issue #${issueNumber}: ${error}`,
      blocks,
    });
  }
}
